<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from kr210.urdf.xacro               | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="kuka_kr210_arm">
  <!-- Link0 -->
  <gazebo reference="base_link">
    <material>Gazebo/Grey</material>
    <mu1>2</mu1>
    <mu2>2</mu2>
  </gazebo>
  <!-- Link1 -->
  <gazebo reference="link_1">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>
  <!-- Link2 -->
  <gazebo reference="link_2">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>
  <!-- Link3 -->
  <gazebo reference="link_3">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>
  <!-- Link4 -->
  <gazebo reference="link_4">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>
  <!-- Link5 -->
  <gazebo reference="link_5">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>
  <!-- Link6 -->
  <gazebo reference="link_6">
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>
  <!-- Gripper -->
  <gazebo reference="gripper_link">
    <material>Gazebo/Grey</material>
    <mu1>0.2</mu1>
    <mu2>0.2</mu2>
  </gazebo>
  <gazebo reference="left_gripper_finger_link">
    <material>Gazebo/Grey</material>
    <kp>1000000.0</kp>
    <kd>1.0</kd>
    <mu1>30.0</mu1>
    <mu2>30.0</mu2>
  </gazebo>
  <gazebo reference="right_gripper_finger_link">
    <material>Gazebo/Grey</material>
    <kp>1000000.0</kp>
    <kd>1.0</kd>
    <mu1>30.0</mu1>
    <mu2>30.0</mu2>
  </gazebo>


  <!-- grasp plugin -->
  <gazebo>
    <plugin filename="libgazebo_grasp_fix.so" name="gazebo_grasp_fix">
      <arm>
        <arm_name>kuka_arm</arm_name>
        <palm_link>link_6</palm_link>
        <gripper_link>right_gripper_finger_link</gripper_link>
        <gripper_link>left_gripper_finger_link</gripper_link>
      </arm>
      <forces_angle_tolerance>100</forces_angle_tolerance>
      <update_rate>4</update_rate>
      <grip_count_threshold>4</grip_count_threshold>
      <max_grip_count>8</max_grip_count>
      <release_tolerance>0.003</release_tolerance>
      <disable_collisions_on_attach>false</disable_collisions_on_attach>
      <contact_topic>__default_topic__</contact_topic>
    </plugin>
  </gazebo>
  <!--Links-->
  <link name="base_footprint"/>
  <link name="base_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="150"/>
      <inertia ixx="60" ixy="0" ixz="0" iyy="70" iyz="0" izz="100"/>
    </inertial>
    <visual>
      <origin rpy="-1.570796327 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/visual/base_link.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.75294 0.75294 0.75294 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="-1.570796327 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/collision/base_link.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="link_1">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.4"/>
      <mass value="138"/>
      <inertia ixx="30" ixy="0" ixz="0" iyy="50" iyz="0" izz="50"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/visual/link_1.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.75294 0.75294 0.75294 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/collision/link_1.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="link_2">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0.44751"/>
      <mass value="95"/>
      <inertia ixx="40" ixy="0" ixz="0" iyy="40" iyz="0" izz="10"/>
    </inertial>
    <visual>
      <origin rpy="-1.570796327 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/visual/link_2.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.75294 0.75294 0.75294 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/collision/link_2.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="link_3">
    <inertial>
      <origin rpy="0 0 0" xyz="0.18842 0.18344 -0.042799"/>
      <mass value="71"/>
      <inertia ixx="1" ixy="0" ixz="0" iyy="10" iyz="0" izz="10"/>
    </inertial>
    <visual>
      <origin rpy="-1.570796327 0 0" xyz="0 -0.184 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/visual/link_3.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.75294 0.75294 0.75294 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 -0.184 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/collision/link_3.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="link_4">
    <inertial>
      <origin rpy="0 0 0" xyz="0.27146 -0.007326 0"/>
      <mass value="17"/>
      <inertia ixx="0.18" ixy="0" ixz="0" iyy="0.55" iyz="0" izz="0.64"/>
    </inertial>
    <visual>
      <origin rpy="-1.570796327 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/visual/link_4.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.75294 0.75294 0.75294 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/collision/link_4.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="link_5">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="7"/>
      <inertia ixx="0.04" ixy="0" ixz="0" iyy="0.068" iyz="0" izz="0.068"/>
    </inertial>
    <visual>
      <origin rpy="-1.570796327 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/visual/link_5.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.75294 0.75294 0.75294 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/collision/link_5.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="link_6">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.5"/>
      <inertia ixx="0.004" ixy="0" ixz="0" iyy="0.003" iyz="0" izz="0.003"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/visual/link_6.dae"/>
      </geometry>
      <material name="">
        <color rgba="0.75294 0.75294 0.75294 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/collision/link_6.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="gripper_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.3"/>
      <inertia ixx="0.0001" ixy="0" ixz="0" iyy="0.0002" iyz="0" izz="0.0001"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/visual/gripper_base.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/visual/gripper_base.dae"/>
      </geometry>
    </collision>
  </link>
  <link name="right_gripper_finger_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.002" ixy="0" ixz="0" iyy="0.0002" iyz="0" izz="0.0002"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/visual/finger_right.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/collision/finger_right_collision.dae"/>
      </geometry>
    </collision>
  </link>
  <link name="left_gripper_finger_link">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.002" ixy="0" ixz="0" iyy="0.0002" iyz="0" izz="0.0002"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/visual/finger_left.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="/home/stmark/robot_arm_ws/src/kuka_kr210_arm/meshes/collision/finger_left_collision.dae"/>
      </geometry>
    </collision>
  </link>
  <joint name="right_gripper_finger_joint" type="prismatic">
    <origin rpy="0 0 0" xyz="0.15 -0.0725 0"/>
    <parent link="gripper_link"/>
    <child link="right_gripper_finger_link"/>
    <axis xyz="0 1 0"/>
    <limit effort="100" lower="-0.01" upper="0.06" velocity="0.05"/>
    <dynamics damping="0.7"/>
  </joint>
  <joint name="left_gripper_finger_joint" type="prismatic">
    <origin rpy="0 0 0" xyz="0.15 0.0725 0"/>
    <parent link="gripper_link"/>
    <child link="left_gripper_finger_link"/>
    <axis xyz="0 -1 0"/>
    <limit effort="100" lower="-0.01" upper="0.06" velocity="0.05"/>
    <dynamics damping="0.7"/>
  </joint>
  <joint name="gripper_joint" type="fixed">
    <parent link="link_6"/>
    <child link="gripper_link"/>
    <origin rpy="0 0 0" xyz="0.11 0 0"/>
    <!--0.087-->
    <axis xyz="0 1 0"/>
  </joint>
  <!-- joints -->
  <joint name="fixed_base_joint" type="fixed">
    <parent link="base_footprint"/>
    <child link="base_link"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <joint name="joint_1" type="revolute">
    <origin rpy="0 0 0" xyz="0 0 0.33"/>
    <parent link="base_link"/>
    <child link="link_1"/>
    <axis xyz="0 0 1"/>
    <limit effort="300" lower="-3.2288592050000005" upper="3.2288592050000005" velocity="2.1467550390000003"/>
  </joint>
  <joint name="joint_2" type="revolute">
    <origin rpy="0 0 0" xyz="0.35 0 0.42"/>
    <parent link="link_1"/>
    <child link="link_2"/>
    <axis xyz="0 1 0"/>
    <limit effort="300" lower="-0.785398185" upper="1.4835299050000001" velocity="2.007128695"/>
  </joint>
  <joint name="joint_3" type="revolute">
    <origin rpy="0 0 0" xyz="0 0 1.25"/>
    <parent link="link_2"/>
    <child link="link_3"/>
    <axis xyz="0 1 0"/>
    <limit effort="300" lower="-3.6651915300000004" upper="1.134464045" velocity="1.954768816"/>
  </joint>
  <joint name="joint_4" type="revolute">
    <origin rpy="0 0 0" xyz="0.96 0 -0.054"/>
    <parent link="link_3"/>
    <child link="link_4"/>
    <axis xyz="1 0 0"/>
    <limit effort="300" lower="-6.10865255" upper="6.10865255" velocity="3.124139447"/>
  </joint>
  <joint name="joint_5" type="revolute">
    <origin rpy="0 0 0" xyz="0.54 0 0"/>
    <parent link="link_4"/>
    <child link="link_5"/>
    <axis xyz="0 1 0"/>
    <limit effort="300" lower="-2.1816616250000003" upper="2.1816616250000003" velocity="3.0019663960000003"/>
  </joint>
  <joint name="joint_6" type="revolute">
    <origin rpy="0 0 0" xyz="0.193 0 0"/>
    <parent link="link_5"/>
    <child link="link_6"/>
    <axis xyz="1 0 0"/>
    <limit effort="300" lower="-6.10865255" upper="6.10865255" velocity="3.8222711670000002"/>
  </joint>



  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <robot_sim_type>gazebo_ros2_control/GazeboSystem</robot_sim_type>
      <parameters>/home/stmark/robot_arm_ws/src/kuka_kr210_arm/config/jtc.yaml</parameters>
    </plugin>
  </gazebo>
  


  <ros2_control name="GazeboSystem" type="system">
    <hardware>
      <plugin>gazebo_ros2_control/GazeboSystem</plugin>
  </hardware>

  <joint name="joint_1">
    <command_interface name="position">
      <param name="min">-3.14</param>
      <param name="max">3.14</param>
    </command_interface>
    <command_interface name="velocity">
      <param name="min">-3.15</param>
      <param name="max">3.15</param>
    </command_interface>
    <state_interface name="position"/>
    <state_interface name="velocity"/>
    <state_interface name="effort"/>
    <param name="initial_position">0.0</param>   
  </joint>
  <joint name="joint_2">
    <command_interface name="position">
      <param name="min">-3.14</param>
      <param name="max">3.14</param>
    </command_interface>
    <command_interface name="velocity">
      <param name="min">-3.15</param>
      <param name="max">3.15</param>
    </command_interface>
    <state_interface name="position"/>
    <state_interface name="velocity"/>
    <state_interface name="effort"/>
    <param name="initial_position">-1.57</param>   
  </joint>
  <joint name="joint_3">
    <command_interface name="position">
      <param name="min">-3.14</param>
      <param name="max">3.14</param>
    </command_interface>
    <command_interface name="velocity">
      <param name="min">-3.15</param>
      <param name="max">3.15</param>
    </command_interface>
    <state_interface name="position"/>
    <state_interface name="velocity"/>
    <state_interface name="effort"/>
    <param name="initial_position">0.0</param>   
  </joint>
  <joint name="joint_4">
    <command_interface name="position">
      <param name="min">-3.14</param>
      <param name="max">3.14</param>
    </command_interface>
    <command_interface name="velocity">
      <param name="min">-3.2</param>
      <param name="max">3.2</param>
    </command_interface>
    <state_interface name="position"/>
    <state_interface name="velocity"/>
    <state_interface name="effort"/>
    <param name="initial_position">-1.57</param>   
  </joint>
  <joint name="joint_5">
    <command_interface name="position">
      <param name="min">-3.14</param>
      <param name="max">3.14</param>
    </command_interface>
    <command_interface name="velocity">
      <param name="min">-3.2</param>
      <param name="max">3.2</param>
    </command_interface>
    <state_interface name="position"/>
    <state_interface name="velocity"/>
    <state_interface name="effort"/>
    <param name="initial_position">0.0</param>   
  </joint>
  <joint name="joint_6">
    <command_interface name="position">
      <param name="min">-3.14</param>
      <param name="max">3.14</param>
    </command_interface>
    <command_interface name="velocity">
      <param name="min">-3.2</param>
      <param name="max">3.2</param>
    </command_interface>
    <state_interface name="position"/>
    <state_interface name="velocity"/>
    <state_interface name="effort"/>
    <param name="initial_position">0.0</param>   
  </joint>

  <joint name="right_gripper_finger_joint">
    <command_interface name="position">
      <param name="min">-3.14</param>
      <param name="max">3.14</param>
    </command_interface>
    <command_interface name="velocity">
      <param name="min">-3.2</param>
      <param name="max">3.2</param>
    </command_interface>
    <state_interface name="position"/>
    <state_interface name="velocity"/>
    <state_interface name="effort"/>
    <param name="initial_position">0.0</param>   
  </joint>

  <joint name="left_gripper_finger_joint">
    <command_interface name="position">
      <param name="min">-3.14</param>
      <param name="max">3.14</param>
    </command_interface>
    <command_interface name="velocity">
      <param name="min">-3.2</param>
      <param name="max">3.2</param>
    </command_interface>
    <state_interface name="position"/>
    <state_interface name="velocity"/>
    <state_interface name="effort"/>
    <param name="initial_position">0.0</param>   
  </joint> 

</ros2_control>

</robot>
